### PR TITLE
executor, memory: detach from parent when executor is closed

### PR DESCRIPTION
**Pull request by zz-jason**
We should remove the corresponding Tracker node from current Tracker Tree once an executor is closed.
Before this PR, for some executors, we may start many child executors of them in a meanwhile, but close them respectively, which will cause the tracked total memory usage inaccurate. 
This PR fixes this problem.

### CODE REVIEW

**Comment by shenli** [comment]
Do not leave the description empty.

File: util/memory/tracker.go [hunk 1]
```diff
@@ -84,16 +84,43 @@ func (t *Tracker) AttachTo(parent *Tracker) {
...
+	for i, child := range t.children {
+		if child != oldChild {
+			continue
+		}
+
+		t.bytesConsumed -= oldChild.BytesConsumed()
+		oldChild.parent = nil
+
+		numChildren := len(t.children)
+		if i < numChildren-1 {
```
Comment (coocood) [inline util/memory/tracker.go]: How about
```
t.children = append(t.children[:i], t.children[i+1:]...)
```

File: util/memory/tracker.go [hunk 2]
```diff
@@ -84,16 +84,43 @@ func (t *Tracker) AttachTo(parent *Tracker) {
...
+func (t *Tracker) remove(oldChild *Tracker) {
+	t.Lock()
+	defer t.Unlock()
+	for i, child := range t.children {
+		if child != oldChild {
+			continue
+		}
+
+		t.bytesConsumed -= oldChild.BytesConsumed()
+		oldChild.parent = nil
```
Comment (coocood) [inline util/memory/tracker.go]: Here may have data race?

Comment (zz-jason) [reply util/memory/tracker.go]: no

**Comment by coocood** [comment]
LGTM

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
When will an executor `build->NextChunk->Close their children many times`?

**Pull request merged**